bob-tools/bin/brt.rs

```diff
@@ -0,0 +1,271 @@
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+struct BlobHeader {
+    magic_byte: u64,
+    version: u32,
+    flags: u64,
+}
+
+impl BlobHeader {
+    fn validate(&self) -> AnyResult<()> {
+        if self.magic_byte != BLOB_MAGIC_BYTE {
+            return Err(ValidationError("blob header magic byte is invalid".to_string()).into());
+        }
+        Ok(())
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq)]
+pub struct Header {
+    magic_byte: u64,
+    key: Vec<u8>,
+    meta_size: u64,
+    data_size: u64,
+    flags: u8,
+    blob_offset: u64,
+    created: u64,
+    data_checksum: u32,
+    header_checksum: u32,
+}
+
+impl Header {
+    fn validate(&self) -> AnyResult<()> {
+        if self.magic_byte != RECORD_MAGIC_BYTE {
+            return Err(ValidationError("record header magic byte is invalid".to_string()).into());
+        }
+        let mut header = self.clone();
+        header.header_checksum = 0;
+        let serialized = bincode::serialize(&header)?;
+        let checksum = crc32(&serialized);
+        if self.header_checksum != checksum {
+            return Err(ValidationError(format!(
+                "wrong header checksum: '{}' != '{}'",
+                self.header_checksum, checksum
+            ))
+            .into());
+        }
+        Ok(())
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, Clone, PartialEq)]
+pub struct Record {
+    header: Header,
+    meta: Vec<u8>,
+    data: Vec<u8>,
+}
+
+impl Debug for Record {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(
+            f,
+            "Record(meta_size={}, data_size={})",
+            self.header.meta_size, self.header.data_size
+        )
+    }
+}
+
+impl Record {
+    fn validate(&self) -> AnyResult<()> {
+        self.header.validate()?;
+        let checksum = crc32(&self.data);
+        if self.header.data_checksum != checksum {
+            return Err(ValidationError(format!(
+                "wrong data checksum: '{}' != '{}'",
+                self.header.data_checksum, checksum
+            ))
+            .into());
+        }
+        Ok(())
+    }
+}
+
+fn main() {
+    if let Err(err) = try_main() {
+        log::error!("Program finished with error: {}", err);
+    }
+}
+
+fn try_main() -> AnyResult<()> {
+    init_logger()?;
+    log::info!("Logger initialized");
+    let settings = Settings::from_matches()?;
+    let mut input = Blob::reader(&settings.input)?;
+    log::info!("Blob reader created");
+    let mut output = Blob::writer(&settings.output)?;
+    log::info!("Blob writer created");
+    let header = input.read_header()?;
+    output.write_header(&header)?;
+    log::info!("Input blob header version: {}", header.version);
+    let mut count = 0;
+    while !input.is_eof() {
+        match input.read_record() {
+            Ok(record) => {
+                output.write_record(&record)?;
```

Comment:
Maybe there should be some checksum validation mid-write, like every N records we check that checksum of written records are the same as the read ones (so we keep at max N records in memory, N can be read from config with default value of 100, for example).

Reply:
Now I check the equality of N last cached and written records.